### PR TITLE
Add the refresh_tokens table and handle revoke refresh tokens

### LINKED ISSUE #221: The refresh token should be revocable
#### Components
This proposal is about the **OAuth2** module of **SteemConnect**. To be more precise - about the `https://steemconnect.com/api/oauth2/token/revoke` endpoint.

#### Proposal
At the moment, there is no way to revoke refresh token via the `https://steemconnect.com/api/oauth2/token/revoke` endpoint. 

Currently, this endpoint revokes the *access token* provided in the *Authorization* header.  I believe it would be really useful for developers to be able to also revoke the *refresh token*.

#### Mockups / Examples
My proposal is to change the current behavior of the `https://steemconnect.com/api/oauth2/token/revoke` endpoint. Here is how it IMO could work:
proposal headers
proposal body

The core concept is that **not** the **access_token** provided in the *Authorization* header is being revoked, but **token** provided in the request's body.

It would enable a developer to implement revoking both **refresh token** and **access token** using the same endpoint.

As you can see, the **headers** section would include:
- *Content-Type: application/x-www-form-urlencoded*
- *Accept: application/json*
- *Authorization: <access_token>*

and **body**:
- *token*, the access_token or refresh_token to revoke
- *token_type_hint*, there developer should specify the type of token (*access_token* or *refresh_token*) provided in the token field

In my opinion, if ***token_type_hint*** was *refresh_token*, SteemConnect **should revoke both** access token and refresh token.

If ***token_type_hint*** was *access_token*, SteemConnect **should only revoke** access token.

For more information check this paper:
- https://tools.ietf.org/pdf/rfc7009.pdf

#### Benefits
With revocable refresh token via the `/api/oauth2/token/revoke` endpoint, a developer would be able to delete all tokens of the user if one didn't need them anymore. My proposal would enable to use `offline` scope more safety.

<br /><hr/><em>Posted on <a href="https://utopian.io/utopian-io/@jakipatryk/refresh-token-is-not-revocable">Utopian.io -  Rewarding Open Source Contributors</a></em><hr/>

### PR DESCRIPTION
**Pull request by gregory-latinier**
For the revoke method test if the token is a refresh token and blacklist it
When requesting a new access token using the refresh token, we now test if it's blacklisted
Fix #221 #257

### CODE REVIEW

**Comment by qanon1111** [comment]
I have a question I have created and waited for a replay to my account creation at chat.golem.network/home
Rocket chat
So I chose to say lost password now email.
Why am I having these issues.
Are you checking my remote logs?
IAM trying to understand why there is so much complexity within the platform. IAM sure singularity net will not have these issues. I follow every platform you are on. I have suggested using ipfs quad 9 and other safe forms of transmit and receive over the web. Can someone tell me don't need to register a Google email to get into this chat?
Qanon111@protonmail
You should know who IAM from previous issues and conversation.
Thank you

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by qanon1111** [comment]
A local webserver with ipfs would solve these issues and allow a node system with encryption and encrypted chat and a backend DDOS and front end blacklist of all who take advantage like using of my internet data that is worth alot when a node times ten is riding my coat tail.

Sent from ProtonMail mobile

-------- Original Message --------

File: db/migrations/create-refresh-tokens.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => queryInterface.createTable('refresh_tokens', {
+    id: {
```
[the remark this answers is not in the document]
Comment (gregory-latinier) [reply db/migrations/create-refresh-tokens.js]: May be in the future we would need to query the table for some reason. It doesn't hurt to have extra information. Those tokens are not expired so I'd rather choose `blacklisted_refresh_tokens`.

File: routes/oauth2.js [hunk 2]
```diff
@@ -47,6 +47,23 @@ router.all('/api/oauth2/authorize', authenticate('user'), async (req, res) => {
...
+    const refreshToken = await req.db.blacklisted_refresh_tokens.findOne({
+      where: {
+        client_id: req.proxy,
+        user: req.user,
+        token: req.token,
+      },
+    });
+    if (refreshToken) {
+      res.status(401).json({
+        error: 'refresh_token_revoked',
```
Comment (bonustrack) [inline routes/oauth2.js]: Can you use `access_denied` error?

File: routes/oauth2.js [hunk 3]
```diff
@@ -61,9 +78,18 @@ router.all('/api/oauth2/token', authenticate(['code', 'refresh']), async (req, r
...
-router.all('/api/oauth2/token/revoke', authenticate('app'), async (req, res) => {
-  await req.db.tokens.destroy({ where: { token: req.token } });
-  res.json({ success: true });
+router.all('/api/oauth2/token/revoke', authenticate(['app', 'refresh']), async (req, res) => {
+  if (req.role === 'refresh') {
+    await req.db.blacklisted_refresh_tokens.create({
+      client_id: req.proxy,
+      user: req.user,
+      token: req.token,
+    });
```
Comment (bonustrack) [inline routes/oauth2.js]: We should catch error

File: routes/oauth2.js [hunk 4]
```diff
@@ -47,6 +47,23 @@ router.all('/api/oauth2/authorize', authenticate('user'), async (req, res) => {
 
 /** Request app access token */
 router.all('/api/oauth2/token', authenticate(['code', 'refresh']), async (req, res) => {
+  // Is the refresh token blacklisted ?
+  if (req.role === 'refresh') {
```
Comment (bonustrack) [inline routes/oauth2.js]: This should be in higher level, on middleware authenticate method.

**Comment by bonustrack** [comment]
We keep this in standby, waiting to decide about the implementation we really want for revoke refresh token.

**Pull request closed without merging**